File: Code/Javascript/Observer/observer.js
class Train {
    constructor() {    
        this.passengers = [];
    }

    board(passenger) {
        passenger.enter(this);
        this.passengers.push(passenger)
    }

    discharge(passenger) {
        this.passengers.splice(this.passengers.indexOf(passenger), 1)
    }

    arrive(station) {
        this.station = station;
        this.announceStation();
    }

    announceStation() {
        for(i = this.passengers.length - 1; i >= 0; i--) {
            this.passengers[i].arrive(this.station);
        }
    }
}

class Passenger {
    constructor() {
        if (this.constructor === Passenger) {
            throw new TypeError('Abstract class "Passenger" cannot be instantiated directly.'); 
        }
    }
    enter(train) {
        this.train = train;
    }
}

class CountingPassenger extends Passenger {
    constructor(count){
        super()
        this.count = count;
    }

    arrive(station) {
        this.count--;
        if (this.count == 0) this.train.discharge(this);
    }
}

class CheckingPassenger extends Passenger {
    constructor(station) {
        super()
        this.station = station;
    }

    arrive(station) {
        if (this.station == station) this.train.discharge(this);
    }
}

stations = ["Station A", "Station B", "Station C", "Station D"];
train = new Train();
train.station = stations[0];

for (i = 1; i < 6; i++) {
    train.board(new CountingPassenger(i))
}

for (i = 0; i < 4; i++) {
    train.board(new CheckingPassenger(stations[i]))
}

j = 0
while(train.passengers.length != 0) {
    j = (j+1)%stations.length;
    train.arrive(stations[j]);
}
console.log("Final Station" + train.station);

File: Code/Go/Observer/observer.go
package main

import (
	"container/list"
	"fmt"
)

type Train struct {
	passengers *list.List
	station    string
}

type Passenger interface {
	enter(train *Train)
	arrive(station string)
}

type PassengerStruct struct {
	train *Train
	Passenger
}

type CountingPassenger struct {
	counter int
	PassengerStruct
}

type CheckingPassenger struct {
	destination string
	PassengerStruct
}

func NewTrain() *Train {
	return &Train{list.New(), ""}
}

func NewCountingPassenger(count int) *CountingPassenger {
	return &CountingPassenger{count, PassengerStruct{}}
}

func NewCheckingPassenger(destination string) *CheckingPassenger {
	return &CheckingPassenger{destination, PassengerStruct{}}
}

func (train *Train) board(passenger Passenger) {
	passenger.enter(train)
	train.passengers.PushFront(passenger)
}

func (train Train) discharge(passenger Passenger) {
	for e := train.passengers.Front(); e != nil; e = e.Next() {
		if e.Value.(Passenger) == passenger {
			train.passengers.Remove(e)
			return
		}
	}
}

func (train *Train) arrive(station string) {
	train.station = station
	train.announceStation()
}

func (train Train) announceStation() {
	next := train.passengers.Back()
	for e := train.passengers.Back(); e != nil; e = next {
		next = e.Prev()
		e.Value.(Passenger).arrive(train.station)
	}
}

func (passenger *PassengerStruct) enter(train *Train) {
	passenger.train = train
}

func (passenger *CountingPassenger) arrive(station string) {
	passenger.counter--
	if passenger.counter == 0 {
		passenger.train.discharge(passenger)
	}
}

func (passenger *CheckingPassenger) arrive(station string) {
	if passenger.destination == station {
		passenger.train.discharge(passenger)
	}
}

func main() {
	var train *Train = NewTrain()

	stations := [4]string{"Station1", "Station2", "Station3", "Station4"}

	for i := 1; i < 6; i++ {
		train.board(NewCountingPassenger(i))
	}

	for i := 0; i < 4; i++ {
		train.board(NewCheckingPassenger(stations[i]))
	}

	j := 0
	for train.passengers.Len() != 0 {
		j = (j + 1) % 4
		train.arrive(stations[j])
	}

	fmt.Printf("Final Station: %s", train.station)
}
